File: scripts/adapters/codex.py
"""OpenAI Codex CLI adapter (Python)."""
from __future__ import annotations
import json, re
from datetime import datetime
from pathlib import Path
from typing import Iterator

_INTERRUPT = re.compile(r"abort|interrupt|cancel|sigint", re.I)


def _flatten(c) -> str:
    if isinstance(c, list):
        return "\n".join(
            (it if isinstance(it, str) else it.get("text") or json.dumps(it, ensure_ascii=False))
            for it in c
        )
    if isinstance(c, str):
        return c
    return json.dumps(c, ensure_ascii=False)


def parse(file_path: str) -> Iterator[dict]:
    text = Path(file_path).read_text(encoding="utf-8", errors="ignore")
    for raw in text.splitlines():
        if not raw.strip():
            continue
        try:
            obj = json.loads(raw)
        except Exception:
            continue

        ts = obj.get("timestamp") or obj.get("ts") or obj.get("created_at") or (datetime.utcnow().isoformat() + "Z")
        role = (obj.get("role") or obj.get("type") or (obj.get("message") or {}).get("role") or "").lower()

        actor, action_type = "system", "system"
        if role in ("user", "user_input"):
            actor, action_type = "user", "prompt"
        elif role in ("assistant", "response"):
            actor, action_type = "ai", "ai_response"
        elif role in ("function_call", "tool_call", "shell_call"):
            actor, action_type = "ai", "tool_call"
        elif role in ("patch", "apply_patch") or obj.get("type") == "apply_patch":
            actor, action_type = "ai", "code_edit"

        content = _flatten(obj.get("content") or obj.get("message") or obj.get("text") or raw)
        if _INTERRUPT.search(content) or obj.get("stop_reason") == "interrupted":
            action_type = "interrupt"

        files = [p for p in (obj.get("file_path"), obj.get("path")) if p]
        meta = {"source": "codex"}
        if files:
            meta["files"] = files

        yield {"ts": ts, "actor": actor, "action_type": action_type, "content": content, "meta": meta}

File: scripts/adapters/codex.js
/**
 * OpenAI Codex CLI session adapter.
 * Codex CLI は ~/.codex/sessions/<id>.jsonl 形式を採用し、各行は
 *   {timestamp, type|role, content|message, ...}
 * 形式の OpenAI Responses API 互換 JSON。
 */
const fs = require('fs');

function* parse(filePath) {
  const text = fs.readFileSync(filePath, 'utf8');
  for (const raw of text.split(/\r?\n/)) {
    if (!raw.trim()) continue;
    let obj;
    try { obj = JSON.parse(raw); } catch { continue; }

    const ts = obj.timestamp || obj.ts || obj.created_at || new Date().toISOString();
    const role = (obj.role || obj.type || obj.message?.role || '').toLowerCase();

    let actor = 'system';
    let action_type = 'system';
    if (role === 'user' || role === 'user_input') {
      actor = 'user';
      action_type = 'prompt';
    } else if (role === 'assistant' || role === 'response') {
      actor = 'ai';
      action_type = 'ai_response';
    } else if (role === 'function_call' || role === 'tool_call' || role === 'shell_call') {
      actor = 'ai';
      action_type = 'tool_call';
    } else if (role === 'patch' || role === 'apply_patch' || obj.type === 'apply_patch') {
      actor = 'ai';
      action_type = 'code_edit';
    }

    let content = obj.content ?? obj.message ?? obj.text ?? raw;
    if (Array.isArray(content)) {
      content = content.map(c => (typeof c === 'string' ? c : c.text || JSON.stringify(c))).join('\n');
    } else if (typeof content !== 'string') {
      content = JSON.stringify(content);
    }

    if (/abort|interrupt|cancel|sigint/i.test(content) || obj.stop_reason === 'interrupted') {
      action_type = 'interrupt';
    }

    const files = [];
    if (obj.file_path) files.push(obj.file_path);
    if (obj.path) files.push(obj.path);

    yield {
      ts,
      actor,
      action_type,
      content,
      meta: { source: 'codex', files: files.length ? files : undefined },
    };
  }
}

module.exports = { parse };
